### ISSUE TITLE
Introduce Versioning of compatibility between agents and contracts

### ISSUE THREAD

**Issue by fredo** [beamer-bridge/beamer#473]
Problem description: 

We are currently working on the contract interfaces. Short term this will be done a couple of times more. At the same time we are fixing functionality or improving performance on the agent.
Just updating the production agent to the latest main will break compatibility with the deployed contracts. At the same time we want to have the fixes being updated. 

Outcome of the discussion:

General:
- Make sure that agents and contracts interface is always compatible for each version
- Create a forbid requests switch
- Upon each new deployment let the old agent run until all its requests/claims are resolved and deploy a new agent in parallel.
- Define a versioning scheme to align with the above

For Testnet: 

- have rapid development cycles
- Deploy any sprint (if necessary ) new contracts and agents

For Mainnet deployment:
- Define a more "long term" interface between contracts and agents

@palango @istankovic please add to this comment in case I'm missing something

**Comment by fredo** [comment]
Outcome of today's discussion:

As a simple way to mitigate versioning problems between agents and contracts we can do the following:

- introduce version numbers for contracts
- Provide abstract interfaces on the agent's side for each contract version
- In the agent's core use abstract functions (withdraw, claim, etc)
- Having another module which handles the interaction with the contract specified by the version

This is applicable to events as well as contract calls

It would be nice to have a way to automatically check whether the interface between the agent and the contract has changed and that we need to upgrade the version. 

If I have forgotten anything important, please add @weilbith

**Comment by fredo** [comment]
I think it is very cheap to introduce now compared to introducing in the future.

**Comment by istankovic** [comment]
One potential issue that I see here is that if the request/claim state machine changes, it's going to be non-trivial to adjust for that, unless the state machine is cloned and modified, which not nice. In other words, how do we accommodate changes to the state machine(s) between versions?

Additionally, this issue seems more like an epic and it would be good to have much smaller, concrete and actionable issues written down as children of this one.

**Comment by weilbith** [comment]
@istankovic Simple: the state machine is part of the core. It is the central and most important business logic we have. In a circular dependency graph it is in the center of the agent. Following dependency management concepts this means this core must depend on nothing else. Only components on other levels of this circle depend on it (from the outside to the inside). This means two things. First of all if the core changes, this can trigger change in all outside layers too, the other way around not. Second, to reduce the impact of the first case, having very generic and stable interfaces between the circles and components is key. For example at a very outside layer of the circle we have a module which is capable of doing an on-chain transaction. It takes care of re-tries, gas, pricing, require errors, etc pp. The whole rest of the agent does not care at all how this is done. In best case, most components (usually increasingly towards the center) the components don't even know what a transaction is and that this involves a blockchain. 
In this case there are three layers for me. The protocol with its state machine, a layer on top which abstracts away the execution actions triggered by the state machine (here the versioning of different contracts comes into play). And finally on the very outside a layer which is capable of interacting with the blockchain and other peripheral systems the agent interacts with.

**Comment by istankovic** [comment]
@weilbith I am not sure that I understand what you are trying to say. To expand on my question:

> In other words, how do we accommodate changes to the state machine(s) between versions?

Contract changes have so far required changes to the state machine and the state machine determine determines the agent's behaviour. I suppose things will stay that way for the foreseeable future.

So let us imagine we built an abstraction within the agent that covers the state machine. Once we change the contracts and we need to update the state machine, what are we to do? We could copy the complete state machine, but that is quite far from ideal, in my opinion. Not to mention that we would need such copies _anytime_ we update the state machine.

Also, it is far from obvious to me how we would prevent abstraction leaks.

**Comment by fredo** [comment]
@istankovic could you maybe give a concrete example of a scenario to expand your worries a bit? Maybe this will clear things up between you and @weilbith

**Comment by istankovic** [comment]
> @istankovic could you maybe give a concrete example of a scenario to expand your worries a bit? Maybe this will clear things up between you and @weilbith

Sure. A perfect example would be the upcoming non-fill proof changes. Both the contracts and agents behaviour will change.
But one needs not look at such a dramatic example. Just adding a new contract event would suffice.
A handler for that event needs to be written, and the state machine needs to consider that event, perhaps there may be a new state or several introduced. Obviously, that would result in a new version of the state machine, so my question is, how are we going to handle such changes?

Note that this is more of a technical question, rather than a conceptual. IOW, I'm trying to see how we might address this on a technical level, preferably without cloning the entire state machine.

**Comment by weilbith** [comment]
@istankovic let us have a discussion in person in the office tomorrow. I think this will be much more productive. We will leave a comment with the outcome here in the thread.
